imap_processing/ialirt/l0/ialirt_spice.py

```diff
@@ -42,69 +45,158 @@ def get_z_axis(sc_inertial_right: NDArray, sc_inertial_decline: NDArray) -> NDAr
 
     Parameters
     ----------
     z_axis : NDArray
-        Unit vector spacecraft Z-axis.
+        Unit vector of spacecraft spin axis (Z-axis).
     spin_phase : NDArray
         Spin phase angle in radians.
 
     Returns
     -------
     rot_matrices : NDArray
-        Rotation matrix.
-
-    Notes
-    -----
-    This matrix acts just like SPICE's pxform(instrument_frame, "IMAP_SPACECRAFT", et).
-    A forward rotation that transforms vectors from the instrument's local frame
-    to the spacecraft’s rotating frame (URF)
+        Rotation matrices to rotate vectors around Z by spin_phase.
     """
-    # Rotation matrix to rotate about z_axis by spin_phase
     rot_matrices = np.array(
         [spice.axisar(z, float(phase)) for z, phase in zip(z_axis, spin_phase)]
     )
 
     return rot_matrices
 
 
-def transform_instrument_vectors_to_urf(
+def get_x_y_axes(z_axis: NDArray) -> NDArray:
+    """
+    Build orthonormal frames from input Z-axis vectors.
+
+    Parameters
+    ----------
+    z_axis : NDArray
+        Array of spacecraft Z-axis unit vectors, shape (N, 3).
+
+    Returns
+    -------
+    frames : NDArray
+        Array of rotation matrices, shape (N, 3, 3).
+    """
+    # Pick a fixed reference vector.
+    v_ref = np.array([0, 0, 1])
+
+    # Detect if z_axis is nearly aligned with v_ref.
+    dot_products = np.dot(z_axis, v_ref)
+    too_parallel = np.abs(dot_products) > 0.99
+
+    # Use alternate reference vector where needed.
+    v_refs = np.tile(v_ref, (z_axis.shape[0], 1))
+    v_refs[too_parallel] = np.array([1, 0, 0])
+
+    # Compute a temporary Y-axis: perpendicular to both z_axis and v_ref.
+    y_temp = np.cross(z_axis, v_refs)
+    # Make it a unit vector.
+    y_axis = y_temp / np.linalg.norm(y_temp, axis=-1, keepdims=True)
+
+    # Take the cross product to get the X-axis.
+    x_axis = np.cross(y_axis, z_axis)
+
+    frames = np.stack([z_axis, y_axis, x_axis], axis=1)
+
+    return frames
+
+
+def compute_total_rotation(
+    inertial_frames: NDArray, spin_rotations: NDArray, mount_matrix: NDArray
+) -> NDArray:
+    """
+    Map instrument vectors to inertial space.
+
+    Parameters
+    ----------
+    inertial_frames : NDArray
+        Spacecraft to inertial rotation matrices (N, 3, 3).
+    spin_rotations : NDArray
+        Spacecraft spin rotation matrices (N, 3, 3).
+    mount_matrix : NDArray
+        Matrix for instrument to spacecraft alignment (3, 3).
+
+    Returns
+    -------
+    total_rotations : NDArray
+        Instrument to inertial rotation matrices (N, 3, 3).
+    """
+    total_rotations = []
+
+    for rotation_sc, spin in zip(inertial_frames, spin_rotations):
+        # Multiply the three matrices: inertial, spin, and mount.
+        # instrument → spacecraft → rotated spacecraft → inertial
+        rotation_inst_to_inertial = rotation_sc @ spin @ mount_matrix
+
+        total_rotations.append(rotation_inst_to_inertial)
+
+    total_rotations = np.array(total_rotations)
```

Comment:
I believe that numpy will correctly to the broadcasting of these arrays such that matrices of sizes `(N, 3, 3) @ (N, 3, 3) @ (3, 3) = (N, 3, 3)`
```suggestion
    for rotation_sc, spin in zip(inertial_frames, spin_rotations):
        # Multiply the three matrices: inertial, spin, and mount.
        # instrument → spacecraft → rotated spacecraft → inertial
        rotation_inst_to_inertial = rotation_sc @ spin @ mount_matrix

        total_rotations.append(rotation_inst_to_inertial)

    total_rotations = rotation_sc @ spin @ mount_matrix
```
From the documentation:

> The behavior depends on the arguments in the following way.
    - If both arguments are 2-D they are multiplied like conventional matrices.
    - If either argument is N-D, N > 2, it is treated as a stack of matrices residing in the last two indexes and broadcast accordingly.

https://numpy.org/doc/2.1/reference/generated/numpy.matmul.html